KAFKA-5886: Implement KIP-91 delivery.timeout.ms

sutambe opened the pull request:
First shot at implementing kip-91 `delivery.timeout.ms`. Summary

1. Added `delivery.timeout.ms` config. Default 120,000
2. Changed `retries` to `MAX_INT`.
3. batches may expire whether they are inflight or not. So muted is no longer used in `RecordAccumulator.expiredBatches`.
4. In some rare situations batch.done may be called twice. Attempted transitions from failed to succeeded are logged. Successful to successful is an error (exception as before). Other transitions (failed-->aborted, aborted-->failed) are ignored.
5. The old test from `RecordAccumulatorTest` is removed. It has three additional tests. `testExpiredBatchSingle`, `testExpiredBatchesSize`, `testExpiredBatchesRetry`. All of them test that expiry is independent of muted.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -98,6 +98,12 @@
+    /** <code>delivery.timeout.ms</code> */
+    public static final String DELIVERY_TIMEOUT_MS_CONFIG = "delivery.timeout.ms";
+    private static final String DELIVERY_TIMEOUT_MS_DOC = "An upper bound on the time to report success or failure to send a message. "
+                                                          + "The time is measured from the point when send returns. "
```
It seems better to say Producer.send() instead of send.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -76,13 +76,13 @@
 
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long now) {
-        this(tp, recordsBuilder, now, false);
+    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long createTime) {
```
We are passing `now` everywhere else. Maybe we can just keep the argument name the same.

sutambe replied:
The actual argument is `now`. However, I like the formal argument name to be `createTime` because it's an immutable value while constructing a batch. `now`, is by definition, changing.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -288,29 +292,19 @@ public String toString() {
-            abortRecordAppends();
-        return expired;
+    boolean maybeExpire(long deliveryTimeoutMs, long now) {
+        if (deliveryTimeoutMs < (now - this.createdMs)) {
```
Should <= be used ?

sutambe replied:
agreed

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -267,34 +267,28 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
-                            break;
-                        }
+            synchronized (dq) {
+                // iterate over the batches and expire them if they have been in the accumulator for more than requestTimeOut
```
deliveryTimeoutMs should be mentioned

apurvam commented:
Thanks for the PR @sutambe . Looking over the changes, it seems that there are two cases from the KIP which don't seem to be covered: 

1. Setting the `pollTimeout` to be the expiryTime of the oldest batch being sent in the produce request. I think we need this to make sure that we expire batches in a timely manner.
2. Related to the previous point, the current patch doesn't seem to expire inflight requests, which is another feature that the KIP seems to promise.

Have I missed something? Or are you planning on adding the functionality above?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -232,6 +238,7 @@
                                 .define(COMPRESSION_TYPE_CONFIG, Type.STRING, "none", Importance.HIGH, COMPRESSION_TYPE_DOC)
                                 .define(BATCH_SIZE_CONFIG, Type.INT, 16384, atLeast(0), Importance.MEDIUM, BATCH_SIZE_DOC)
                                 .define(LINGER_MS_CONFIG, Type.LONG, 0, atLeast(0L), Importance.MEDIUM, LINGER_MS_DOC)
+                                .define(DELIVERY_TIMEOUT_MS_CONFIG, Type.LONG, 120 * 1000, atLeast(0L), Importance.MEDIUM, DELIVERY_TIMEOUT_MS_DOC)
```
Should we validate the delivery.timeout.ms is greater than request.timeout.ms?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -164,25 +164,29 @@ public void abort(RuntimeException exception) {
     public void done(long baseOffset, long logAppendTime, RuntimeException exception) {
-        final FinalState finalState;
-        if (exception == null) {
+        final FinalState tryFinalState = (exception == null) ? FinalState.SUCCEEDED : FinalState.FAILED;
```
It is probably cleaner to have an explicit `EXPIRED` state.

sutambe replied:
I did some digging around. An expired batch's final state is `FAILED`. I don't feel great about adding yet another `finalState`. We already have `ABORTED` and `FAILED`. The `ProducerBatch.done` will get even more complicated.

becketqin replied:
Maybe it's not a big deal but just want to call out that this is a behavior change. Currently the producer will throw exception when transition from FAILED state to another state due to some reason other than expiration. If we change this logic, we may miss those cases which are not failed by expiration but still got state update twice. It may not be that important if we do not have programming bugs.

Personally I think it is better to clearly define the states of the batches even if additional complexity is necessary.

The comments should probably also cover the force close case for completeness.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -164,25 +164,29 @@ public void abort(RuntimeException exception) {
-            if (this.finalState.get() == FinalState.ABORTED) {
-                log.debug("ProduceResponse returned for {} after batch had already been aborted.", topicPartition);
-                return;
+        if (!this.finalState.compareAndSet(null, tryFinalState)) {
```
The logic here probably needs more comments. We may have the following three cases that the state of a batch has been updated before the ProduceResponse returns:
1. A transaction abortion happens. The state of the batches would have been updated to `ABORTED`.
2. The producer is closed forcefully. The state of the batches would have been updated to `ABORTED`.
3. The batch is expired when it is in-flight. The state of the batch would have been updated to `EXPIRED`.

In the other cases, we should throw IllegalStateException.

sutambe replied:
Please review the updated method documentation.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -267,34 +267,28 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
-                            // Stop at the first batch that has not expired.
-                            break;
-                        }
+            synchronized (dq) {
```
The batches still needs to be expired in order if `max.in.flight.requests.per.connection` is set to 1. So we probably still want to check if the partition is muted or not. That said, if we guarantee that when `RecordAccumulator.expiredBatches()` returns non-empty list, all the earlier batches have already been expired, we can remove the muted check here.

BTW, I did not see the logic of expiring an in-flight batch in the current patch. Am I missing something?

sutambe replied:
It's there now

sutambe replied:
@becketqin Added the muted check back

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -267,34 +267,28 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
+                Iterator<ProducerBatch> batchIterator = dq.iterator();
+                while (batchIterator.hasNext()) {
+                    ProducerBatch batch = batchIterator.next();
+                    boolean isFull = batch != lastBatch || batch.isFull();
```
isFull is no longer used.

becketqin replied:
Still not used.

becketqin reviewed (commented):
Thanks for the patch. Left some comments.

apurvam commented:
Heads up @sutambe , the following PR just got merged, and may have some conflicts with the current patch : https://github.com/apache/kafka/pull/3743

There shouldn't be any impact on the logic, however.

ijuma commented:
Friendly reminder that the feature freeze is this Wednesday.

becketqin commented:
@ijuma Just want to check. Do you think this feature is a "minor" feature?

ijuma commented:
@becketqin, it is possible to classify this as a minor feature, but the fact that it affects a core part of the Producer puts it in a bit of a grey area. If the PR is almost ready and we miss the feature freeze, my take is that it would be OK to merge it by the end of this week. Later than that and it seems a bit risky.

It's a bit worrying that the merge conflicts haven't been resolved since last week.

sutambe commented:
@ijuma  @becketqin I've an new PR but after a rebase I've to fix one more test. Working on that now.

ijuma commented:
Thanks @sutambe!

sutambe commented:
@apurvam It's not clear to me why `testExpiryOfFirstBatchShouldCauseResetIfFutureBatchesFail` and `testExpiryOfFirstBatchShouldNotCauseUnresolvedSequencesIfFutureBatchesSucceed` are failing. It looks like a batch that should be in IncompleteBatches isn't there. Any thoughts?

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -271,7 +276,8 @@ private long sendProducerData(long now) {
 
-        List<ProducerBatch> expiredBatches = this.accumulator.expiredBatches(this.requestTimeout, now);
+        List<ProducerBatch> expiredBatches = this.accumulator.expiredBatches(this.deliveryTimeoutMs, now);
+        boolean needsTransactionStateReset = false;
```
This variable can be dropped.

tedyu commented:
I added the following to testExpiryOfFirstBatchShouldCauseResetIfFutureBatchesFail before the first sender.run() call
```
        Sender sender = new Sender(logContext, client, metadata, this.accumulator, true, MAX_REQUEST_SIZE, ACKS_ALL, 10,
            new Metrics(), new SenderMetricsRegistry(), time, REQUEST_TIMEOUT, DELIVERY_TIMEOUT, 50, transactionManager, apiVersions);
```
The test still fails.

apurvam commented:
@sutambe where are those tests failing? The latest PR builder suggests that the clients and core tests all passed.

sutambe commented:
@apurvam @ijuma @becketqin The Sender and RecordAccumulator are passing now. The failing tests are connect tests that are irrelevant.

apurvam commented:
@sutambe I don't think the test failures are irrelevant since the same 3 tests failed in all the runs. Further, the cause of the failure is: 

```
java.lang.AssertionError: 
  Unexpected method call Listener.onFailure(job-0, org.apache.kafka.common.KafkaException: Failed to construct kafka producer):
```

I think their mocks may need to be updated to take account of the new configs and attendant `ConfigExceptions`

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,41 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
         List<ProducerBatch> expiredBatches = new ArrayList<>();
+
+        // Expiration of inflight batches happens here in the order of createdMs if the final state of the
+        // batch in not known by (batch's creation time + deliveryTimeoutMs).
```
in not -> is not

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,41 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
+            Map.Entry<Long, List<ProducerBatch>> expiredInflightBatches = inFlightBatches.pollFirstEntry();
+            for (ProducerBatch batch: expiredInflightBatches.getValue()) {
+                if (batch.getFinalState() == null &&
+                    batch.maybeExpire(deliveryTimeoutMs, now)) {
```
The check 'if (deliveryTimeoutMs <= (now - this.createdMs))' inside maybeExpire() would be true.
Looks like another method can be created inside ProducerBatch which expires the batch.

sutambe replied:
`maybeExpire` has a side-effect of setting `errorMessage` internally. Hence calling it again in `if`.

tedyu replied:
Understand.
That part can be refactored - goal is to reduce unnecessary comparison.

sutambe replied:
@apurvam Those test don't even compile or run on my machine. What's up with those tests?

apurvam replied:
They can't construct a kafka producer with the changes made in this PR.

becketqin replied:
Assuming `nFlightBatches` is a TreeSet suggested above, this code can be simplified to:
```
        while (!inFlightBatches.isEmpty() &&
               inFlightBatches.first().maybeExpire(deliveryTimeoutMs, now)) {
            expiredBatches.add(inFlightBatches.pollFirst());
        }
```

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,32 +158,47 @@ public void abort(RuntimeException exception) {
+                }
+                // No-op for transitions from ABORTED-->FAILED and FAILED-->ABORTED
             } else {
+                // A SUCCESSFUL batch must not succeed again.
```
Is this comment accurate? The new state is not necessarily SUCCEEDED.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -298,29 +313,20 @@ public String toString() {
-     *     <li> the batch is not in retry AND request timeout has elapsed after it is ready (full or linger.ms has reached).
-     *     <li> the batch is in retry AND request timeout has elapsed after the backoff period ended.
-     * </ol>
+     * Expire the batch if it no outcome is known in within delivery.timeout.ms.
```
Some typos in this comments. "Expire the batch if no outcome is known within delivery.timeout.ms"

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -85,6 +90,10 @@
 
+    // A mapping of batch's creation time to the batch for quick of the oldest batch
+    // and lookup in the order of creation
+    private final NavigableMap<Long, List<ProducerBatch>> inFlightBatches;
```
Does this have to be a per partition Map? Intuitively we just need a `TreeSet<ProducerBatch>` with a comparator?

becketqin replied:
Apparently the my understanding of `TreeSet` is not accurate. It uses the comparator to decide whether the entries are the same or not. We can use a TreeMap<Long, Set<ProducerBatch>> then. We may also want to bucket the timestamp a little bit, say one second to avoid huge amount of Sets created for each ms in the `TreeMap`.

tedyu replied:
I was thinking about this too. Using millisecond as unit for Map key is not prudent.

After the switch to second as unit, we may need to check the two adjacent buckets keyed by ts-1  (sec) and ts+1 (sec).

sutambe replied:
As we discussed, `TreeSet` does not cut it. The naming is consistent. A `TreeSet` is a set. It's just that equality criterion is different.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,41 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
+
         for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches.entrySet()) {
             Deque<ProducerBatch> dq = entry.getValue();
             TopicPartition tp = entry.getKey();
```
`tp` is not used anymore.

hachikuji replied:
Still not used

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,41 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
-                        }
+            synchronized (dq) {
+                // iterate over the batches and expire them if they have been in the accumulator for more than deliveryTimeoutMs
+                ProducerBatch lastBatch = dq.peekLast();
```
No longer used.

hachikuji replied:
This is still not used

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -304,10 +326,21 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
      */
     public void reenqueue(ProducerBatch batch, long now) {
         batch.reenqueued(now);
+        List<ProducerBatch> inflight = inFlightBatches.get(batch.createdMs);
```
This logic would become `inFlightRequests.remove(batch)` when a `TreeSet` is used for this.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -572,6 +616,16 @@ public boolean hasUndrained() {
                                         batch.close();
                                         size += batch.records().sizeInBytes();
                                         ready.add(batch);
+                                        // Put this batch in the list of inflight batches because we might have
```
This would be just `inFlightBatches.add(batch)`

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -586,6 +640,10 @@ public boolean hasUndrained() {
         return batches;
     }
 
+    public Long getEarliestDeliveryTimeoutMs() {
```
We usually just use `earliestDeliveryTimeout` in Kafka.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -132,6 +135,7 @@ public Sender(LogContext logContext,
                   SenderMetricsRegistry metricsRegistry,
                   Time time,
                   int requestTimeout,
+                  long deliveryTimeoutMs,
```
It seems we don't need the `deliveryTimeoutMs` in the sender. It is only used as an argument passed to the accumulator. But the accumulator already has the config.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -271,7 +276,7 @@ private long sendProducerData(long now) {
         }
 
-        List<ProducerBatch> expiredBatches = this.accumulator.expiredBatches(this.requestTimeout, now);
+        List<ProducerBatch> expiredBatches = this.accumulator.expiredBatches(this.deliveryTimeoutMs, now);
```
It seems an existing issue. When we expire the batches here. The memory of those batches will be deallocated. It seems that we will deallocate the same batch again when the ProduceResponse returns?

becketqin reviewed (commented):
@sutambe Thanks for updating the patch. Made a pass on the non-test file and left some comments. Will review the tests tomorrow. We may need to have some quick turnaround to get this into 1.0.0.

apurvam commented:
@sutambe I had a look at the failing Sender expiry tests. What is happening is that the tests are not modified to account for the fact that the inflight batches can be expired. In the tests, we used to expire a batch sitting in the accumulator but _not_ the inflight batch. When the inflight batch returned, it would be re queued. 

But now, the test sends the response for the inflight batch, but when it goes to requeue, it discovers that there shouldn't be an inflight request an raises an exception.

The tests should be updated to account for the new behavior and make sure that the inflight batch is _not_ expired.

apurvam commented:
Actually, the test reveals a bug in the current patch: the response for the inflight batch which expired is not being handled correctly. We should not be trying to requeue it to start with.

So we need two tests: one where the inflight batch is not expired, and the current case. The reenqueue logic in the sender needs to be updated to not reenqueue the expired batches.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java:
```diff
@@ -214,40 +214,27 @@ public void testSplitPreservesMagicAndCompressionType() {
+     * as not expired by {@link ProducerBatch#maybeExpire(long, long)}.
      */
     @Test
     public void testLargeLingerOldNowExpire() {
```
This test has nothing to do with linger.ms anymore...

becketqin replied:
We should change the test name to something like testBatchExpiration. and the test below to testBatchExpirationAfterReenqueue.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java:
```diff
@@ -214,40 +214,27 @@ public void testSplitPreservesMagicAndCompressionType() {
+     * as not expired by {@link ProducerBatch#maybeExpire(long, long)}.
      */
     @Test
     public void testLargeRetryBackoffOldNowExpire() {
```
Similar to above we should rename this.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -332,10 +332,13 @@ public void testNextReadyCheckDelay() throws Exception {
+        long lingerMs = Integer.MAX_VALUE / 16;
+        long retryBackoffMs = Integer.MAX_VALUE / 8;
+        int requestTimeoutMs = Integer.MAX_VALUE / 4;
+        long deliveryTimoutMs = Integer.MAX_VALUE;
```
typo: timeout

becketqin replied:
The typo is still there.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -510,76 +513,125 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
+    public void testExpiredBatchSingle() throws InterruptedException {
         long lingerMs = 3000L;
-        int requestTimeout = 60;
+        long deliveryTimoutMs = 3200L;
```
typo: timeout

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,42 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
+                // If the final state of the batch is known (success or failed) just skip past it because
+                // the callback has already been invoked. This allows the clean up of the map in bulk right here.
+                if (batch.finalState() == null) {
+                    batch.expire(now);
```
Should we still expire the batches when they are expired instead of expiring all the bucket? Having a second granularity bucket does not prevent us from doing that, right?

becketqin reviewed (commented):
@sutambe Thanks for updating the batch. A few comments:
1. for a batch that is got expired prematurely, we should not reqenqueu the batch. (as @apurvam noticed) and we should not double deallocate the memory.
2. There are a few review comments before that are not addressed yet. (such as unused local variables)
3. We may want to revisit some of the tests and see if they still make sense.
4. It would be good to add more unit tests to the patch. More specifically, we may want to have the following tests:
* Test a batch is correctly inserted into the in.flight.batches if needed. And not inserted if not needed.
* Test the callback of an expired batch is fired in time when it is in-flight/not in-flight
* Test when expire an in-flight batch, we still wait for the request to finish before sending the next batch.
* Test we are not going to retry an already expired batch.
* Test when batch is expired prematurely, the buffer pool is only deallocated after the response is returned. (because we are still holding the batch until the response is returned)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -435,8 +438,23 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+            throw new ConfigException("Must set " + ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG + " higher than " +
+                                      ProducerConfig.LINGER_MS_CONFIG + ", " +
+                                      ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG + ", and " +
+                                      ProducerConfig.RETRY_BACKOFF_MS_CONFIG + " combined.");
```
Hmm.. Might not be too important, but it doesn't seem necessary to include the retry backoff in this check. If the user sets retries=0, then the backoff shouldn't matter.

hachikuji replied:
Please respond to this. Is it necessary to include RETRY_BACKOFF in this check?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -298,29 +313,27 @@ public String toString() {
-        else if (this.inRetry() && requestTimeoutMs < (waitedTimeMs(now) - retryBackoffMs))
-            expiryErrorMessage = (waitedTimeMs(now) - retryBackoffMs) + " ms has passed since last attempt plus backoff time";
+    boolean maybeExpire(long deliveryTimeoutMs, long now) {
+        if (deliveryTimeoutMs <= (now - this.createdMs)) {
```
We are using the creation time of the batch to check for expiration. That will tend to expire some records which were added to the batch after creation earlier than the delivery timeout (by as much as linger.ms). Alternatively, we could use the time that the batch was closed, which will tend to expire records later than the delivery timeout (by as much as linger.ms), but maybe expiring late is bit safer than expiring early? This is equivalent to saying that the delivery timeout excludes linger time.

sutambe replied:
Batch close may be arbitrrily delayed in some cases. See @junrao's explanation: http://search-hadoop.com/m/Kafka/uyzND1calgR1Udv2J?subj=Re+DISCUSS+KIP+91+Provide+Intuitive+User+Timeouts+in+The+Producer

hachikuji replied:
Thanks. I synced with Jun and it seems reasonable. It would help to document somewhere why we use create time.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -298,29 +313,27 @@ public String toString() {
+     */
+    void expire(long now) {
+        expiryErrorMessage = (now - this.createdMs) + " ms has passed since batch creation";
+        abortRecordAppends();
```
Checking my understanding. With this change, it should no longer be possible to expire a batch before linger.ms has completed and the batch has been closed. If so, do we still need the logic to abort appends on expiration? (It might be safer to have it anyway, just checking if it is still needed for correctness)

hachikuji replied:
I think the answer to this question is that it is possible to expire while the batch is still being built because closing the batch can be arbitrarily delayed by inflight fetches.

sutambe replied:
Right

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -494,6 +529,13 @@ public boolean hasUndrained() {
+        // By the end of this function, we'll bring it down as we look at the batches.
+        // This ensures that earliest delivery timeout keeps moving ahead.
+        if (now >= earliestDeliveryTimeoutMs) {
+            earliestDeliveryTimeoutMs = Long.MAX_VALUE;
```
After we reset `earliestDeliveryTimeoutMs`, it seems that we do not take into account the expiration times of in-flight batches.

sutambe replied:
fixed. take a look.

sutambe replied:
`RecordAccumulator.maybeUpdateEarliestDeliveryTimeout`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -586,6 +642,21 @@ public boolean hasUndrained() {
 
+    /**
+     * Bucket the batch into a bucket that's one second wide only for lookup in the
+     * inflightBatches map. This is an optimization to reduce the number of keys in the
```
Do we have any microbenchmarks that show this (potential) optimization is justified?

hachikuji replied:
If you are not implementing this, can you please remove it?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -572,6 +618,16 @@ public boolean hasUndrained() {
                                         ready.add(batch);
+                                        // Put this batch in the list of inflight batches because we might have
+                                        // to expire it if we don't know its final state by deliveryTimeoutMs
+                                        if (batch.createdMs + deliveryTimeoutMs <= now + requestTimeoutMs) {
```
Why do we need this check? A comment would be helpful.

hachikuji reviewed (commented):
Thanks for the patch. Left a few comments/questions.

sutambe commented:
@becketqin  Added a couple of tests. 
Test a batch is correctly inserted into the in.flight.batches if needed. And not inserted if not needed.
`RecordAccumulatorTest.testSoontoExpireBatchesArePickedUp`
Test the callback of an expired batch is fired in time when it is in-flight/not in-flight
`SenderTest.testInflightBatchesExpireOnDeliveryTimeout`

hachikuji commented:
@sutambe I don't see any changes since my comments. Are you sure you have pushed them?

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -317,6 +342,16 @@ public void reenqueue(ProducerBatch batch, long now) {
         }
     }
 
+    private void maybeRemoveFromInflight(ProducerBatch batch) {
```
Shouldn't we also be removing the batches from the inflight set when the batch is completed (failed or successfully)? I might be missing something, but I don't see that code here.

sutambe replied:
@apurvam Right. Cleanup of `soonToExpireInFlightBatches` happens in two places (1) if a batch gets reenqueued and (2) when `Sender` looks for `expiredBatches`. In the second case, we cleanup the batches whose final state is known (success or failure) and there by "removing" them.

apurvam replied:
Hmm. This seems a bit off. What this means that in the 'normal' case when responses are successful and there is no backlog in the accumulator, we will hang on to batches (and not garbage collect them) until the delivery timeout. 

Indeed, if you add the following at the end of sender tests where there are no more inflight requests (ie. all requests have completed and will never be retried) `        assertTrue(accumulator.soonToExpireInFlightBatches().isEmpty());` all the tests fail.

I think this should be fixed. We should clear the `soonToExpireInflightBatches` as soon as the batch is completely resolved (ie. failed or completed) so that we don't hang on to the reference unnecessarily.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -149,41 +149,57 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
+    public boolean abort(RuntimeException exception) {
         if (!finalState.compareAndSet(null, FinalState.ABORTED))
-            throw new IllegalStateException("Batch has already been completed in final state " + finalState.get());
+            //throw new IllegalStateException("Batch has already been completed in final state " + finalState.get());
```
As far as I can tell, it shouldn't be possible to abort a batch after it has been completed. Is this correct? If so, I think it might be better to continue to raise `IllegalStateException`. It's preferable to keep the allowable state transitions as narrowly defined as possible since it ensures faster failure for unexpected paths.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -85,6 +90,10 @@
     private final TransactionManager transactionManager;
 
+    // A mapping of batch's creation time to the batch for quick of the oldest batch
+    // and lookup in the order of creation
```
nit: "for quick **access** to the oldest batch"?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -75,6 +77,9 @@
     private final long retryBackoffMs;
+    private final int requestTimeoutMs;
+    private final long deliveryTimeoutMs;
+    private long earliestDeliveryTimeoutMs;
```
The name is a little misleading given its proximity to similarly named fields. Maybe something like `nextExpirationTimestampMs` would be clearer?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -98,6 +98,11 @@
+    /** <code>delivery.timeout.ms</code> */
+    public static final String DELIVERY_TIMEOUT_MS_CONFIG = "delivery.timeout.ms";
+    private static final String DELIVERY_TIMEOUT_MS_DOC = "An upper bound on the time to report success or failure after Producer.send() returns. "
+                                                          + "Producer may report failure to send a message earlier than this config if all the retries are exhausted";
```
Do we want to mention the other case where a record is expired early because it was added to a batch which was already nearing expiration?

sutambe replied:
@hachikuji I think `retryBackOff` can be dropped. Perhaps we can do two tests based on whether `retries` is set or not.

becketqin replied:
@sutambe I also think we should mention the scenario that a Record is added to a batch that is about to expire.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -511,6 +548,10 @@ public boolean hasUndrained() {
                                 boolean backoff = first.attempts() > 0 && first.waitedTimeMs(now) < retryBackoffMs;
+                                // We determine the earliest time any batch may expire. We use
+                                // this value later on to determine the max time we can sleep in poll.
+                                earliestDeliveryTimeoutMs = Math.min(earliestDeliveryTimeoutMs, first.createdMs + deliveryTimeoutMs);
```
The logic for updating this field seems to assume that the batch at the front of the deque will always be the next to expire, but I'm not sure that is true in the case of retries.

becketqin replied:
Good point. It would work if max.in.flight.requests.per.connection=1, or when we enable idempotence. But otherwise the first batch may not be the first to expire. Since we do not even have callback order guarantee in that case, maybe it is fine? But we should definitely document this.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,42 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
+
+        // Expiration of inflight batches happens here in the order of createdMs if the final state of the
+        // batch is not known by (batch's creation time + deliveryTimeoutMs).
+        while (!soonToExpireInFlightBatches.isEmpty() &&
```
To be honest, this lazy expiration seems like overkill. It should be a rare case where we actually have entries in `soonToExpireInFlightBatches` because of the other optimization to only add to it when the delivery timeout will expire prior to the request timeout. And if the producer is in a situation where batches are being expired, then the performance of removal for a particular batch is probably not a major concern. Maybe some benchmarking would show whether it is a worthwhile optimization.

becketqin replied:
@hachikuji I agree in most case we probably do not need this. It is probably only useful for large producers (e.g. mirror maker which has thousands of partitions to send). There may still be value to have this:

1. Since delivery.timeout.ms and request.timeout.ms are both user configurations. It would be good to guard against some configurations. (e.g. request.timeout.ms=delivery.timeout.ms, linger.ms=0). 

2. It seems that in some scenarios this would help. For example, when the brokers are being rolling bounced, there will be a lot of retried batches. Those batches may have `remainingDeliveryTimeoutMs` < `request.timeout.ms`. We may have even more than one batches per partition to insert if max.in.flight.request is greater than 1.

A benchmark would be useful. But in general I think it is safer to have this optimization given it does not increase too much complexity.

hachikuji commented on clients/src/test/resources/log4j.properties:
```diff
@@ -18,4 +18,4 @@ log4j.appender.stdout=org.apache.log4j.ConsoleAppender
 log4j.appender.stdout.layout.ConversionPattern=[%d] %p %m (%c:%L)%n
 
-log4j.logger.org.apache.kafka=ERROR
+log4j.logger.org.apache.kafka=TRACE
```
Inadvertent commit I assume.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1674,16 +1674,48 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+
+        Map<TopicPartition, ProduceResponse.PartitionResponse> responseMap = new HashMap<>();
+        responseMap.put(tp0, new ProduceResponse.PartitionResponse(Errors.NONE, 0L, 0L, 0L));
+        client.respond(new MockClient.RequestMatcher() {
```
If we're just returning `true` for `matches`, we don't need to provide a `RequestMatcher` at all.

hachikuji reviewed (commented):
A few more comments. Note that there are some failing test cases as well.

sutambe commented:
@becketqin  Added more tests
Test when expire an in-flight batch, we still wait for the request to finish before sending the next batch = `SenderTest.testWhenFirstBatchExpireNoSendSecondBatchIfGuranteeOrder`
Test we are not going to retry an already expired batch = `SenderTest.testExpiredBatchDoesNotRetry`

Regarding the last test "Test when batch is expired prematurely, the buffer pool is only deallocated after the response is returned. (because we are still holding the batch until the response is returned)". I'm not sure if it's really needed because `RecordAccumulator.abortBatches` deallocates all incomplete batches (whether they have been sent or not).

becketqin commented:
@sutambe I am looking at the update patch.

> Regarding the last test "Test when batch is expired prematurely, the buffer pool is only deallocated after the response is returned. (because we are still holding the batch until the response is returned)". I'm not sure if it's really needed because RecordAccumulator.abortBatches deallocates all incomplete batches (whether they have been sent or not).

`RecordAccumulator.abortBatches()` is only used when the producer is force closed. We do not need to worry about the memory consumption at that point anymore because the producer has stopped accepting new messages.

becketqin commented:
@sutambe It looks there are 3 test failures from connect. Can you check?

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -585,6 +639,21 @@ public boolean hasUndrained() {
+     * @return
+     */
+    static private long bucketedCreationTime(ProducerBatch batch) {
+        return batch.createdMs;
```
I don't see bucketing

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -268,34 +283,42 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
+                // If the final state of the batch is known (success or failed) just skip past it because
+                // the callback has already been invoked. This allows the clean up of the map in bulk right here.
+                if (batch.finalState() == null) {
+                    batch.expire(now);
```
I had a comment earlier that it seems we should still expire the batch at exact createTime + deliveryTimeoutMs even if we bucket them by seconds.

sutambe replied:
@becketqin The batch is expired at exact time but not removed from the `soonToExpireInflightBatches`. Based on some earlier comments grouping them to avoid pointer chasing?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -438,8 +441,32 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+        long retryBackoffMs = config.getLong(ProducerConfig.RETRY_BACKOFF_MS_CONFIG);
+        int retries = config.getInt(ProducerConfig.RETRIES_CONFIG);
+
+        if (retries > 0 &&
```
I agree with @hachikuji that using retry backoff is a little weird here. `retries` means that "at most retry that many times within delivery.timeout.ms". So even if `retries` is greater than 0, it does not mean there must be a retry. So we should probably just check the `delivery.tiemout.ms` is at least `linger.ms` + `request.timeout.ms`.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -149,41 +149,59 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
      * @param exception The exception to use to complete the future and awaiting callbacks.
      */
-    public void abort(RuntimeException exception) {
+    public boolean abort(RuntimeException exception) {
```
Can we update the Java doc to explain the return value?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -149,41 +149,59 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
-        final FinalState finalState;
-        if (exception == null) {
+    public boolean done(long baseOffset, long logAppendTime, RuntimeException exception) {
+        final FinalState tryFinalState = (exception == null) ? FinalState.SUCCEEDED : FinalState.FAILED;
```
Personally I still think a clear EXPIRED state would be clearer. We can let batch.done() method take a FinalState argument instead of inferring the state from the exception.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java:
```diff
@@ -72,7 +72,11 @@ public void testBatchAbort() throws Exception {
         // subsequent completion should be ignored
         batch.done(500L, 2342342341L, null);
-        batch.done(-1, -1, new KafkaException());
+        try {
```
This test will pass whether line 76 throws IllegalStateException or not. Should we add a fail() statement after line 76?

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java:
```diff
@@ -72,7 +72,11 @@ public void testBatchAbort() throws Exception {
+        try {
+            batch.done(-1, -1, new KafkaException());
+        } catch (IllegalStateException e) {
+            assertNotNull(e);
```
Is this needed? In what case could e be null?

sutambe replied:
@becketqin restored the test as it was before.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -510,76 +513,156 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
     @Test
-    public void testExpiredBatches() throws InterruptedException {
-        long retryBackoffMs = 100L;
+    public void testSoontoExpireBatchesArePickedUp() throws InterruptedException {
```
testSoonToExpire... (Upper case "To")

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -510,76 +513,156 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
+        assertTrue(drained.isEmpty());
+        assertTrue(accum.soonToExpireInFlightBatches().isEmpty());
+
+        // advanced clock and send one batch out but it should be included in soon to expire inflight
```
Do you mean it should NOT be included...

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1704,16 +1714,96 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+    }
+
+    @Test
+    public void testWhenFirstBatchExpireNoSendSecondBatchIfGuranteeOrder() throws InterruptedException {
```
typo in the test name.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1704,16 +1714,96 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+        setupWithoutTransactionState(true);
+
+        // Send first ProduceRequest
+        Future<RecordMetadata> request1 = accumulator.append(tp0, time.milliseconds(), "key".getBytes(), "value".getBytes(), null, null, MAX_BLOCK_TIMEOUT).future;
```
request1 and request 2 are not used.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1704,16 +1714,96 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+        setupWithoutTransactionState(false);
+
+        // Send first ProduceRequest
+        Future<RecordMetadata> request1 =
```
Should we check the completeness of request1?

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1704,16 +1714,96 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+            }
+        }, produceResponse(tp0, -1, Errors.NOT_LEADER_FOR_PARTITION, -1)); // return a retriable error
+
+        sender.run(time.milliseconds()); // receive first response
```
We may need to call sender.run() one more time to ensure the message is not reenqueued. The reqenqueued message won't be sent out again in the same sender.run().

becketqin reviewed (commented):
@sutambe Thanks for the update. Made another pass and left some comments.

becketqin commented:
@sutambe BTW, I did not see the tests to ensure we are not double deallocating memory. Will you add those tests?

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -77,13 +77,13 @@
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long now, boolean isSplitBatch) {
-        this.createdMs = now;
-        this.lastAttemptMs = now;
+    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long createTime, boolean isSplitBatch) {
```
createTime -> creationTime

sutambe replied:
@becketqin Just added a test to ensure that we don't do double deallocation. `SenderTest.testNoDoubleDeallocation`

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -149,41 +149,59 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
+        if (this.finalState.compareAndSet(null, tryFinalState)) {
+            completeFutureAndFireCallbacks(baseOffset, logAppendTime, exception);
+            return true;
+        } else {
```
nit: 'else' can be dropped

apurvam commented:
FYI, the connect tests are still failing with:

```
Caused by: org.apache.kafka.common.config.ConfigException: Must set delivery.timeout.ms higher than linger.ms + request.timeout.ms
	at org.apache.kafka.clients.producer.KafkaProducer.configureDeliveryTimeout(KafkaProducer.java:452)
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:380)
	... 35 more
```

It would be good to get some green builds before this can be considered safe for merge.

sutambe commented:
@apurvam @becketqin Fixed Connect tests.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -145,45 +145,66 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
+     * Try aborting the batch and complete the future and callbacks. A batch may have already FAILED
+     * due to delivery timeout expiry. In such a case, it simply returns false.
      * @param exception The exception to use to complete the future and awaiting callbacks.
+     * @return Return true if the batch was actually aborted, false otherwise.
```
This method either throw exception or return true, which indicates there is no need to have a return value.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -146,44 +146,61 @@ private boolean tryAppendForSplit(long timestamp, ByteBuffer key, ByteBuffer val
      * @param exception The exception to use to complete the future and awaiting callbacks.
      */
-    public void abort(RuntimeException exception) {
+    public boolean abort(RuntimeException exception) {
```
The method never returns false. We can keep it as void if so.

sutambe replied:
@becketqin return type restored to `void`

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/ProducerBatchTest.java:
```diff
@@ -214,40 +214,27 @@ public void testSplitPreservesMagicAndCompressionType() {
         ProducerBatch batch = new ProducerBatch(new TopicPartition("topic", 1), memoryRecordsBuilder, now);
         // Set `now` to 2ms before the create time.
-        assertFalse(batch.maybeExpire(10240, 100L, now - 2L, Long.MAX_VALUE, false));
+        assertFalse(batch.maybeExpire(10240, now - 2L));
```
Should probably add an expire case?

sutambe replied:
added

apurvam commented:
I have a high level comment about the inflight batch handling. I was wondering what the advantages/disadvantages of the current approach (where batches are grouped by creation time) would be against adding a `Map<TopicPartition, PriorityQueue<ProducerBatch>>` to the RecordAccumulator.  This would complement the existing `Map<TopicPartition, Deque<ProducerBatch>>`. 

The priority queue would be ordered by the batch creation time (similar to the current `NavigableMap`).

We would add to the queue on drain. And we would remove from the queue in `Sender.failBatch` and `Sender.completeBatch`, just like we do for the `TransactionManager` (see: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java#L589). So we don't have any special logic for 

The `RecordAccumulator.expiredBatches` would now iterate over both maps (`TopicPartition -> Deque` and `TopicPartition -> PriorityQueue`), with the priority queue being iterated on first. This doesn't change the time complexity of the algorithm, since we are iterating over all partitions in `expiredBatches` once anyway.

The advantage of this approach is that managing inflight batches is much simpler (add to the queue on drain, remove on completion). It doesn't complicate the expiry in any way.

The main disadvantage that I see is in computing the `earliestDeliveryTimeout`. For this you would need to peek at the head of the queue for each partition which has an inflight batch. But this should be fairly cheap, since we would be iterating over 1000's of objects even in the largest installations.

What do you think?

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1036,6 +1043,65 @@ public void testExpiryOfFirstBatchShouldNotCauseUnresolvedSequencesIfFutureBatch
+        assertTrue(request1.isDone());
+        assertEquals(1, client.inFlightRequestCount());
+
+        Map<TopicPartition, ProduceResponse.PartitionResponse> responseMap = new HashMap<>();
```
The map is not used.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1036,6 +1043,65 @@ public void testExpiryOfFirstBatchShouldNotCauseUnresolvedSequencesIfFutureBatch
+        sender.run(time.milliseconds());  // expire the batch
+        assertTrue(request1.isDone());
+        assertEquals(1, client.inFlightRequestCount());
+
```
Should we assert the pool is not deallocated after the expiration but before the response returns?

sutambe replied:
The new code seems to deallocate the batch right away. I'm not changing the behavior for now.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1704,16 +1775,92 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+        time.sleep(deliveryTimeoutMs); // expire the batch
+        sender.run(time.milliseconds());
+
+        // Send second ProduceRequest
```
How could the order be violated if we only append the first batch after the first one is expired?

sutambe replied:
Fixed the test

apurvam commented:
retest this please

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1704,16 +1775,92 @@ public void testTransactionalSplitBatchAndSend() throws Exception {
+        assertTrue(request1.isDone());
+        assertEquals(1, client.inFlightRequestCount());
+
+        Map<TopicPartition, ProduceResponse.PartitionResponse> responseMap = new HashMap<>();
```
The map is not used.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1837,17 +1984,30 @@ private ProduceResponse produceResponse(TopicPartition tp, long offset, Errors e
         this.metrics = new Metrics(metricConfig, time);
-        this.accumulator = new RecordAccumulator(logContext, batchSize, 1024 * 1024, CompressionType.NONE, 0L, 0L, metrics, time,
-                apiVersions, transactionManager);
+        if (customPool == null)
```
nit: Can we avoid reusing the argument name?

becketqin reviewed (commented):
Thanks for updating the patch. I think it is closed to be merged. Just a few more comments.

apurvam commented:
I made this comment in a thread, but that may be buried because github collapses comments. There is a bug in the current implementation which means that we will retain references to `ProducerBatches` in the producer long after they have been successfully completed. The following sequence of assertions in any `SenderTest` shows the presence of this bug: 

```java
    assertFalse(client.hasInFlightRequests());
    // the following should be true, since there are no in flight requests.
    assertFalse(accumulator.soonToExpireInFlightBatches().isEmpty());
    time.sleep(130000);
    // we now 'expire' the completed inflight batches
    sender.run(time.milliseconds());
    // the following shows that we hang on to references for 120 seconds by default.
    assertTrue(accumulator.soonToExpireInFlightBatches().isEmpty());
```

becketqin commented:
@apurvam Initially I was under the same impression. One concern of the priority queue is that whenever we need to remove things out of the queue, we will have to iterate over it. Since a producer can poll thousands of times per seconds, and maybe remove tens of thousands of batches. That said, in most cases, the queue would be empty because most of the batches are sent when remaining delivery.timeout.ms is greater than request.timeout.ms. However, it may still slowdown things when retries occur. My take on this is that if the implementation complexity does not increase much, we may want to follow the robust way instead.

Regarding the memory issue. We could not release the memory of an expired batch right away when the batch is still in-flight. This is because the in-flight batch is still referred by the RequestCompletionHandler of the request. We could remove it from the RequestCompletionHandler, but this is a little tricky.

apurvam commented:
@becketqin I am not sure I follow this: 

> Regarding the memory issue. We could not release the memory of an expired batch right away when the batch is still in-flight. This is because the in-flight batch is still referred by the RequestCompletionHandler of the request. We could remove it from the RequestCompletionHandler, but this is a little tricky.

Is this a response to my comment above where I shared a snippet from `SenderTest` to show that we are retaining batches after they are completed? If so, I didn't mean that we need to release an inflight batch right away. What I was pointing out is that the _only_ way the current patch is deleting batches is by expiring them. This may not be a big deal, since the deallocation is done when the batch is completed, but it still seems like an unnecessary anomaly.

apurvam commented:
Also, it looks like recent commits are causing threads to leak, @sutambe .

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -625,8 +628,9 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
-        batch.done(baseOffset, logAppendTime, exception);
-        this.accumulator.deallocate(batch);
+        if (batch.done(baseOffset, logAppendTime, exception)) {
+            this.accumulator.deallocate(batch);
```
Actually, it seems we may release the memory before the the response returns?

apurvam replied:
As I mentioned in my other comment, the memory utilization itself is not the issue as much as the fact that we are retaining the reference to `ProducerBatch` for at least `deliveryTimeoutMs`. This probably won't cause too much memory pressure, but is still an undesirable behavior.

becketqin replied:
Got it. Good catch. Yes, we should remove the completed batch.

sutambe replied:
@becketqin @apurvam Can some please clarify what suggestion is made here? Remove completed batch from where?

apurvam commented:
> One concern of the priority queue is that whenever we need to remove things out of the queue, we will have to iterate over it. Since a producer can poll thousands of times per seconds, and maybe remove tens of thousands of batches. That said, in most cases, the queue would be empty because most of the batches are sent when remaining delivery.timeout.ms is greater than request.timeout.ms. However, it may still slowdown things when retries occur. My take on this is that if the implementation complexity does not increase much, we may want to follow the robust way instead.

Not sure I follow. It would be no worse that what we do today with the `Deque` except it would be limited only to partitions which had in flight batches. So it is not making things much worse.

becketqin commented:
@apurvam 
> Not sure I follow. It would be no worse that what we do today with the Deque except it would be limited only to partitions which had in flight batches. So it is not making things much worse.

I misunderstood your proposal. I though you were suggesting to use a priority queue to store all the in-flight batches. A `Map<TopicPartition, PriorityQueue<ProducerBatch>>` sounds reasonable.

apurvam commented:
@becketqin 

> I misunderstood your proposal. I though you were suggesting to use a priority queue to store all the in-flight batches. A Map<TopicPartition, PriorityQueue<ProducerBatch>> sounds reasonable.

Yes, I meant a separate `Map<TopicPartition, PriorityQueue<ProducerBatch>>`, to replace the current `NavigableMap`.

hachikuji commented:
@sutambe Any updates on this? Do you need help addressing the test failures?

sutambe commented:
@hachikuji  Sorry for disappearing on you guys. I got sucked into an oncall issue. So I had to context switch. I plan to resume work on this PR next week. I've not had a chance to respond to the comments so far. Thank for you the review so far. It's been very useful.

asfgit commented:
FAILURE
 7505 tests run, 2 skipped, 168 failed.
--none--

asfgit commented:
FAILURE
 7910 tests run, 5 skipped, 48 failed.
--none--

asfgit commented:
FAILURE
 7916 tests run, 5 skipped, 46 failed.
--none--

asfgit commented:
FAILURE
 7280 tests run, 1 skipped, 0 failed.
--none--

asfgit commented:
FAILURE
 7964 tests run, 5 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 7964 tests run, 5 skipped, 1 failed.
--none--

sutambe commented:
@becketqin @apurvam I've updated the PR. Please take a look

asfgit commented:
FAILURE
 7964 tests run, 5 skipped, 17 failed.
--none--

asfgit commented:
FAILURE
 7964 tests run, 5 skipped, 16 failed.
--none--

apurvam commented:
Please fix the failing tests, there is a config exception due to the new delivery timeout config.

asfgit commented:
FAILURE
 7280 tests run, 1 skipped, 0 failed.
--none--

asfgit commented:
FAILURE
 4026 tests run, 1 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 4026 tests run, 1 skipped, 1 failed.
--none--

asfgit commented:
FAILURE
 4026 tests run, 1 skipped, 1 failed.
--none--

asfgit commented:
SUCCESS 
 8077 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8077 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8077 tests run, 5 skipped, 0 failed.
--none--

sutambe commented:
@becketqin @apurvam All checks passed

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -436,6 +441,24 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
         }
     }
 
+    private static long configureDeliveryTimeout(ProducerConfig config) {
```
`verifyAndGetDeliveryTimeout()`?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -436,6 +441,24 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+        int requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+
+        // If lingerMs is Long.MAX_VALUE the sum with requestTimeoutMs overflows
+        if (deliveryTimeoutMs == Long.MAX_VALUE && lingerMs > 0 && requestTimeoutMs > 0 && lingerMs + requestTimeoutMs < 0)
```
This check is a little flaky. What if deliveryTimeoutMs is `Long.MAX_VALUE - 1`?

becketqin reviewed (commented):
Talked to @sutambe and we agreed that @apurvam had a point that we can just use the same way we track accumulated batches. That seems make the code cleaner. Will wait for Sumant to update patch to implement that.

sutambe committed:
Rebasing KIP-91 delivery.timeout.ms for kafka 1.1.0
Avoiding overflow when deliveryTimeoutMs is MAX_VALUE
per-partition map for tracking soon to expire batches
Updated tests

sutambe commented:
@apurvam @becketqin I updated the implementation to use `ConcurrentMap<TopicPartition, Deque<ProducerBatch>>`. Please take a look. I don't see the above test failures on my machine.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -85,12 +89,21 @@
     private final TransactionManager transactionManager;
 
+    // A per-partition queue of batches ordered by creation time for quick access of the oldest batch
+    private final ConcurrentMap<TopicPartition, PriorityQueue<ProducerBatch>> soonToExpireInFlightBatches;
```
We don't need a PriorityQueue for this because the batches in the RecordAccumulator is already in order. So we just need to keep the draining order.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -317,6 +343,13 @@ public void reenqueue(ProducerBatch batch, long now) {
+    private void maybeRemoveFromSoonToExpire(ProducerBatch batch) {
+        PriorityQueue<ProducerBatch> soonToExpireQueue = soonToExpireInFlightBatches.get(batch.topicPartition);
+        if (soonToExpireQueue != null) {
+            soonToExpireQueue.remove(batch);
```
If we always insert the batch to the inFlightBatches queue and there is no bug, the batch to be removed should always be the first batch. Can we assert on that?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -572,6 +614,20 @@ public boolean hasUndrained() {
+                                        // If the expiry is farther than requestTimeoutMs, we don't have to keep
+                                        // track of this batch because it will either succeed or fail (due to
+                                        // request timeout) much sooner.
+                                        if (batch.soonToExpire(deliveryTimeoutMs, now, requestTimeoutMs)) {
```
The original reason we have this optimization is because we used to have a big sorted data structure. So avoiding inserting elements to it makes sense. Given that now the batch order in the RecordAccumulator is already guaranteed. It seems we can just put all the drained batches to the inFlightBatches queue, which is simpler.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -511,6 +544,15 @@ public boolean hasUndrained() {
+                                // We assume that the batch at the front of the deque will always be the next to expire.
+                                // This may not be true if max.in.flight.requests.per.connection > 1 and retries happen.
+                                // Watch for overflow in createdMs + deliveryTimeoutMs when deliveryTimeoutMs is Long.MAX_VALUE
+                                nextBatchExpiryTimeMs = (first.createdMs + deliveryTimeoutMs < 0) ? nextBatchExpiryTimeMs
```
The while loop may break if the request size has reached. So there is no guarantee that it will iterate over all the partitions. One alternative is to find the nextBatchExpiryTimeMs in the expireBatches.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -583,9 +639,39 @@ public boolean hasUndrained() {
+    void maybeUpdateNextExpiryTime(long now) {
+        if (now >= nextBatchExpiryTimeMs) {
+            if (soonToExpireInFlightBatches.isEmpty()) {
+                nextBatchExpiryTimeMs = Long.MAX_VALUE;
```
It seems intuitively this should be the earliest batch in the entire record accumulator?

becketqin commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -277,7 +277,7 @@ private long sendProducerData(long now) {
         }
 
-        List<ProducerBatch> expiredBatches = this.accumulator.expiredBatches(this.requestTimeout, now);
+        List<ProducerBatch> expiredBatches = this.accumulator.expiredBatches(now);
```
It seems we may release the memory for the expired batches before the response is returned. This means the underneath ByteBuffer is still referred by the ProducerBatch instance in the inFlightRequests. I am not sure if this would cause any problem, but it seems a little dangerous.

becketqin commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1866,6 +1882,145 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
+        sender.run(time.milliseconds());  // send request
+        assertEquals(1, client.inFlightRequestCount());
+
+        Map<TopicPartition, ProduceResponse.PartitionResponse> responseMap = new HashMap<>();
```
Is the response preparation needed in this case?

becketqin reviewed (commented):
Thanks for updating the patch. Left some comments.

apurvam commented:
retest this please

apurvam commented:
retest this please

apurvam commented:
So the `org.apache.kafka.clients.producer.internals.SenderTest.testMetadataTopicExpiry` test has failed twice in a row with: 

```
java.lang.ArrayIndexOutOfBoundsException
	at java.base/java.util.zip.CRC32C.update(CRC32C.java:151)
	at org.apache.kafka.common.utils.Checksums.update(Checksums.java:42)
	at org.apache.kafka.common.utils.Crc32C.compute(Crc32C.java:72)
	at org.apache.kafka.common.record.DefaultRecordBatch.writeHeader(DefaultRecordBatch.java:468)
	at org.apache.kafka.common.record.MemoryRecordsBuilder.writeDefaultBatchHeader(MemoryRecordsBuilder.java:357)
	at org.apache.kafka.common.record.MemoryRecordsBuilder.close(MemoryRecordsBuilder.java:311)
	at org.apache.kafka.clients.producer.internals.ProducerBatch.close(ProducerBatch.java:427)
	at org.apache.kafka.clients.producer.internals.RecordAccumulator.drain(RecordAccumulator.java:614)
	at org.apache.kafka.clients.producer.internals.Sender.sendProducerData(Sender.java:270)
	at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:238)
	at org.apache.kafka.clients.producer.internals.SenderTest.testMetadataTopicExpiry(SenderTest.java:473)
```

Given that these changes are on the same code, and given the consistent failure of this test, it is probably a regression. @sutambe can you reproduce the failure locally?

apurvam commented:
Just looking at the stack trace and the test, it may be that an expired batch is being closed twice in some cases.

hachikuji commented:
@sutambe @becketqin It would be nice to unblock this. Can someone else pick up the work?

becketqin commented:
@hachikuji Yeah, this has been pending for too long. I have spoken to @sutambe and he said he still wants to finish the patch. He will figure out the ETA and see if that works.

bbejeck commented:
@sutambe @becketqin is there any update on the status of this PR?  It would be great if we could get this in the next release.

Ishiihara commented:
@guozhangwang @junrao @bbejeck @becketqin We also hit this issue when running Kafka Streams library with some high volume output topics. It would be nice to get this moving and push it to the next release.

radai-rosenblatt commented:
becket cant load this page for some reason (some weird issue with his github profile?).
we are ok with you taking over this patch.

sutambe commented:
@bbejeck @apurvam @becketqin @hachikuji I don't have any update since Dec last year. Sorry, the work has stalled and it has been very hard to find cycles for this effort. I don't mind if Confluent wants to take this effort forward. Better later than never.

Ishiihara commented:
cc @abbccdda @yuyang08

yuyang08 commented:
@sutambe   i made some change based on your pull request to fix style check and test failure. do yo mind I amend the change to this pull request?  cc @becketqin  @apurvam @hachikuji  
  https://github.com/yuyang08/kafka/commit/69fc79a91d0556408c8037649f1e03aa56206ef2

guozhangwang commented:
@yuyang08 I'd suggest you create your own PR against apache kafka trunk and let other reviewers to continue reviewing that one.

yuyang08 commented:
@guozhangwang  sure.  will create a separate pull request

yuyang08 commented:
@guozhangwang  @apurvam @becketqin  created new pr https://github.com/apache/kafka/pull/5270 for KAFKA-5886

ijuma commented:
This has been merged via a different PR, closing.
